### DISCUSSION TITLE
Transfer function

### DISCUSSION THREAD

**Discussion by NjegosJankovic** [CURENT/andes#379] [Q&A] [answered]
Can the transfer function be obtained between two points in the system? I looked through documentation and examples, but I have not found anything on this topic. What I have in mind is the way to select two points in the system, and get the numerical representation of the system between these two points in the transfer function. I assume that the symbolical expression would be more challenging to obtain and less useful.

**Comment by cuihantao** [comment, marked as answer]
I'm not sure what you mean by "numerical representation of the system in the transfer functions". Are you looking to observe the numerical output for a given set of inputs?

The software is not like Simulink, which is causal. It's a non-causal numerical solution of DAE models with discontinuities, so I'm sure if theoretically the transfer functions betwen two points be obtained.

**Reply by NjegosJankovic** [reply]
Thank you for your quick reply. Sorry for not describing the question clearly enough. What I had in mind is a system frequency response. Let's say, for a certain operating point the system frequency response is measured from one signal in the system to the other. I know that something similar exists in PowerFactory. There, the result is frequency response in form of a Bode plot.
In [this discussion ](https://github.com/cuihantao/andes/discussions/153#discussion-3428523) I saw your comment about obtaining the B matrix manually. What about C and D matrices? If these can be obtained also, I can calculate the frequency response.

**Reply by cuihantao** [reply]
Thank you for the clarification. I understand you are looking to obtain a bode plot from the model. The current program is not structured in a way that represents a dynamic power network as an input-output system. When we look at the whole system, it is autonomous, or say self-contained, with no input or output. That's why the state matrix `A` is for the whole system, with no mention of B, C or D.

The `y` specific remains to be defined for the subsystem of interest. Note that the whole differential-algebraic equations defines algebraic variables with a coincidental name `y`, but they are not necessarily a linear combination of `x` and `u`. Voltage at a bus is an algebraic variable, but there is no explicit solution for it. The DAE's `y` also includes the controller outputs, which can be linear. Their equations can be used to find out C by expressig the outputs as a function of `x`. 

To obtain the Bode plot between selected points, some nontrivial work is needed to define the I/O for the selected subsystem, extract the partial derivatives `Fx`, `Fy`, `Gx`, and `Gy` from the Jacobian matrix, and reconstruct the ABCD matrices.

**Reply by NjegosJankovic** [reply]
Thank you very much for your quick and detailed response. I understand your point and I agree that such a requirement is not common and not needed in the majority of power system studies. At the moment I might go to another tool with such an option, as I need it relatively soon. Nonetheless, I might come back to the Andes in future. I really like the idea and the way it is presented and structured.

**Reply by cuihantao** [reply]
I just called a friend in control studies. He corrected me that B, C, and D matrices can be defined based on the need to control and observe the system. I need to walk back from my previous argument; there is no need to manipulate the Jacobian matrices. Since defining the input and outputs will not change the system characteristics (the linearization), the A matrix should stays unchanged. 

So using the TGOV1 model as an example (https://docs.andes.app/en/latest/modeling/examples.html#tgov1), we want to observe the output `Pout` following an input `\Delta Pref` to the leftmost summation block. We can manually substitute the second and third algebraic equation into the first differential equation to find out the coefficient for the input that goes to `B`. Of course, the indices for the `x_{LG}` differential variable needs to be retrieved to construct `B`.

In terms of the output, we will need to solve for `P_{out}` as a function of the states. This can be trivial for LTI systems. Oftentimes, a symbolic toolbox can help.

So in all, this can be done with relatively small efforts. `A` can be obtained from `calc_As()` function in the eigenvalue routine.

**Reply by NjegosJankovic** [reply]
This procedure does seem feasible. I will definitely give it a go.
Thank you very much for all the answers.
